# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
cd data

C:\Users\admin\Downloads\1588204778_GuidedCapstoneStep4_and_step5HL.ipynb\data


In [3]:
df = pd.read_csv('step3_output.csv')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
dfo=df.select_dtypes(include=['object'])

In [5]:
dfo.head(2)

Name    state
0     Hilltop Ski Area   Alaska
1  Sunrise Park Resort  Arizona

In [6]:
df = pd.concat([df.drop('state',axis=1), pd.get_dummies(dfo['state'])], axis=1)

In [7]:
df.head(4)

Unnamed: 0  Unnamed: 0.1                           Name  summit_elev  \
0           2             2               Hilltop Ski Area         2090   
1           4             4            Sunrise Park Resort        11100   
2           5             5  Yosemite Ski & Snowboard Area         7800   
3           8             8         Boreal Mountain Resort         7700   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0            294      0        0.0          0          0     0  ...   
1           1800      0        0.0          0          1     2  ...   
2            600      0        0.0          0          0     0  ...   
3            500      0        0.0          0          1     1  ...   

   Rhode Island  South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0             0          0     0        0         0           0   
1             0             0          0     0        0         0           0   
2             0             0          0     0        0         0           0   
3             0             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  

[4 rows x 62 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [8]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [9]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [10]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [11]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [12]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [13]:
explained_variance_score(y_test,y_pred)

-6.811377891503056e+24

In [14]:
mean_absolute_error(y_test,y_pred)

5387610286738.305

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [15]:
lm.intercept_

-1346902571627.411

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [16]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])


Coefficient
Unnamed: 0         3.194655e+00
Unnamed: 0.1       2.953430e+00
summit_elev        3.191657e+00
vertical_drop      2.698406e+00
trams              1.104340e+14
fastEight          1.093099e+14
fastSixes          1.087862e+14
fastQuads          2.594246e+13
quad               3.103902e+13
triple             5.451136e+13
double             6.673648e+13
surface            6.386554e+13
total_chairs       1.215632e+14
Runs               8.984375e-01
TerrainParks       1.751953e+00
LongestRun_mi      1.910156e+00
SkiableTerrain_ac  1.626953e+00
Snow Making_ac     5.390625e-01
daysOpenLastYear   2.015625e+00
yearsOpen          2.343750e-01
averageSnowfall    1.546875e+00
AdultWeekday       8.972656e+00
projectedDaysOpen  2.667969e+00
NightSkiing_ac     2.343750e-01
clusters           1.539062e+00
Alaska             3.382628e+12
Arizona            3.382628e+12
California         1.134352e+13
Colorado           7.476843e+12
Connecticut        7.476843e+12
Idaho              8.166487e+12
Illinois           5.825308e+12
Indiana            4.770071e+12
Iowa               5.825308e+12
Maine              7.476843e+12
Maryland           3.382628e+12
Massachusetts      6.707017e+12
Michigan           1.256590e+13
Minnesota          9.374213e+12
Missouri           4.770071e+12
Montana            6.707017e+12
Nevada             3.382628e+12
New Hampshire      1.089365e+13
New Jersey         2.120047e+13
New Mexico         8.166487e+12
New York           1.428277e+13
North Carolina     5.825308e+12
Ohio               6.707017e+12
Oregon             5.825308e+12
Pennsylvania       1.089365e+13
Rhode Island       3.382628e+12
South Dakota       3.382628e+12
Tennessee          3.382628e+12
Utah               5.825308e+12
Vermont            9.374213e+12
Virginia           4.770071e+12
Washington         6.707017e+12
West Virginia      5.825308e+12
Wisconsin          1.089365e+13
Wyoming            7.476843e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [17]:
df.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'Name', 'summit_elev', 'vertical_drop',
       'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters',
       'Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Tennessee', 'Utah', 'Vermont', 'Virginia',
       'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype='object')

In [18]:
X1=df.drop(['Name','AdultWeekend','Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Tennessee', 'Utah', 'Vermont', 'Virginia',
       'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'], axis=1)

In [19]:
scaler1 = StandardScaler().fit(X1)

In [20]:
X1_scaled = scaler1.transform(X1)

In [21]:
X1_train, X1_test, y1_train, y1_test = train_test_split(X1_scaled, y, test_size=0.25, random_state=1)

In [22]:
lm1 = linear_model.LinearRegression()

In [23]:
model1 = lm1.fit(X1_train,y1_train)

In [24]:
y1_pred = model1.predict(X1_test)

In [25]:
pd.DataFrame(abs(lm1.coef_), X1.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0         4.140152e-02
Unnamed: 0.1       4.140152e-02
summit_elev        2.811042e+00
vertical_drop      2.228074e+00
trams              8.881784e-16
fastEight          1.332268e-15
fastSixes          0.000000e+00
fastQuads          3.480157e-01
quad               2.334148e+00
triple             4.428025e-01
double             5.775762e-01
surface            1.170665e+00
total_chairs       5.708640e-01
Runs               7.825806e-01
TerrainParks       2.407650e+00
LongestRun_mi      1.623265e+00
SkiableTerrain_ac  1.632050e+00
Snow Making_ac     6.479211e-01
daysOpenLastYear   2.382965e+00
yearsOpen          4.244934e-01
averageSnowfall    9.537911e-01
AdultWeekday       9.671370e+00
projectedDaysOpen  2.151731e+00
NightSkiing_ac     2.728334e-01
clusters           3.292538e+00

In [26]:
explained_variance_score(y1_test,y1_pred)

0.5427563460360543

In [27]:
mean_absolute_error(y1_test,y1_pred)

7.2376162385637395

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [28]:
X2=df.drop(['Name','AdultWeekend','Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Tennessee', 'Utah', 'Vermont', 'Virginia',
       'Washington', 'West Virginia', 'Wisconsin', 'Wyoming','summit_elev','vertical_drop'], axis=1)

In [29]:
scaler2 = StandardScaler().fit(X2)

In [30]:
X2_scaled = scaler2.transform(X2)

In [31]:
X2_train, X2_test, y2_train, y2_test = train_test_split(X2_scaled, y, test_size=0.25, random_state=1)

In [32]:
lm2 = linear_model.LinearRegression()

In [33]:
model2 = lm2.fit(X2_train,y2_train)

In [34]:
y2_pred = model2.predict(X2_test)

In [35]:
pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0         8.154608e-02
Unnamed: 0.1       8.154608e-02
trams              1.776357e-15
fastEight          8.881784e-16
fastSixes          8.881784e-16
fastQuads          7.188539e-01
quad               2.167855e+00
triple             3.931465e-01
double             6.708774e-01
surface            1.094154e+00
total_chairs       6.766942e-01
Runs               1.279520e+00
TerrainParks       2.176139e+00
LongestRun_mi      2.013803e+00
SkiableTerrain_ac  2.044414e+00
Snow Making_ac     9.480265e-01
daysOpenLastYear   2.093504e+00
yearsOpen          2.883744e-01
averageSnowfall    1.607624e+00
AdultWeekday       9.872004e+00
projectedDaysOpen  1.830599e+00
NightSkiing_ac     4.759393e-02
clusters           3.045407e+00

In [36]:
mean_absolute_error(y2_test,y2_pred)

7.1285736332915235

In [37]:
explained_variance_score(y2_test,y2_pred)

0.5513503668926405

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

Model Selection:

# Model 1 (Features Dropped - None)

Explained Variance = -6.81
Mean Absolute Error = 5.38761E+12 


    

# Model 2 (Features Dropped - 'state')
Explained Variance = 0.54
Mean Absolute Error = 7.23 



# Model 3 (Features Dropped - 'state','summit_elev','vertical_drop)
Explained Variance = 0.55
Mean Absolute Error = 7.12


Model 3 & Model 2 both have nearly same values of Mean Error & Variance Score but we will not prefer Model 2 as it incorporates those feature variables which are unimportant to us. 